### PR TITLE
fix boolean check on iterable dataset when len not defined

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes  #6680

`bool(obj)` checks `obj.__bool__` and if it is not defined it will look for `obj.__len__ > 0`. But in case of IterableDataset, neither of these are defined by default, so we need to check for `None` or the empty list explicitly.

**Test fails on master.**

TODO: add to cherry after merge #6083 

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -928,7 +928,7 @@ def test(
         self.testing = True
 
         # If you supply a datamodule you can't supply test_dataloaders
-        if test_dataloaders and datamodule:
+        if test_dataloaders is not None and datamodule:
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:931]: should we apply the same checks to the other methods? e.g. in L869 for `validate` ?

Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:931]: yes, I need to find all occurences

File: tests/trainer/test_dataloaders.py [hunk 2]
```diff
@@ -639,26 +639,37 @@ def test_warning_with_iterable_dataset_and_len(tmpdir):
...
 
+    # without __len__ defined
+    dataloader = DataLoader(IterableWithoutLen(), batch_size=16)
+    assert not has_len(dataloader)
+    assert has_iterable_dataset(dataloader)
+    trainer = Trainer(default_root_dir=tmpdir, max_steps=3)
+    trainer.validate(model, val_dataloaders=dataloader)
+    trainer.fit(model, train_dataloader=dataloader, val_dataloaders=[dataloader])
+    trainer.test(model, test_dataloaders=dataloader)
+
```
Comment (ananthsub) [inline tests/trainer/test_dataloaders.py:672]: Also add a test for predict

**Review by ananthsub — APPROVED** [review]

File: tests/trainer/test_dataloaders.py (hunk 2 above)
Comment (awaelchli) [reply tests/trainer/test_dataloaders.py:675]: done :)

**Review by kaushikb11 — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !
